lib/impl/refactorRules.js

```diff
@@ -0,0 +1,384 @@
+
+        isFieldStillValidTarget(sourceId, formPages, function(invalidField) {
+          if (invalidField) {
+            invalidFields[sourceId] = true;
+          }
+          ruleCb();
+        });
+      }, callback);
+    });
+  }
+
+  function checkRulesWithRepeatingSections() {
+    // Map field ids to actual fields
+    var fieldIdMapping = {};
+    form.pages.forEach(function(page) {
+      page.fields.forEach(function(field) {
+        fieldIdMapping[field._id.toString()] = field;
+      });
+    });
+
+    // Map field ids to it's sections
+    var fieldSectionMapping = getSectionMap(form);
+    fieldSectionMapping = _.mapObject(fieldSectionMapping, function(sectionId) {
+      return fieldIdMapping[sectionId];
+    });
+
+    checkFieldRulesWithRepeatingSections();
+    checkPageRulesWithRepeatingSections();
+
+    /**
+     * Get ids of repeating sections with fields which act as sources in a rule
+     * @param rule
+     * @returns {String[]}
+     */
+    function getSourceRepeatingSections(rule) {
+      return _.chain(rule.ruleConditionalStatements)
+        .map(function(ruleConditionalStatement) {
+          var sourceId = ruleConditionalStatement.sourceField.toString();
+          return fieldSectionMapping[sourceId];
+        })
+        .filter(function(section) {
+          return section && section.repeating;
+        })
+        .map(function(repeatingSection) {
+          return repeatingSection._id.toString();
+        })
+        .uniq()
+        .value();
+    }
+
+    function checkPageRulesWithRepeatingSections() {
+      form.pageRules.forEach(function(pageRule) {
+        var sectionIds = getSourceRepeatingSections(pageRule);
+
+        if (sectionIds.length > 0) {
+          // there is source field in repeating section
+          rulesFlaggedForDeletion[pageRule._id.toString()] = true;
+        }
+      });
+    }
+
+    function checkFieldRulesWithRepeatingSections() {
+      form.fieldRules.forEach(function(fieldRule) {
+        var sectionIds = getSourceRepeatingSections(fieldRule);
+
+        if (sectionIds.length > 1) {
+          // there are more source fields in different repeating sections
+          rulesFlaggedForDeletion[fieldRule._id.toString()] = true;
+        } else if (sectionIds.length === 1) {
+          // all source fields within repeating section are in the same section
+          // -> check that all target fields are in that same section
+          var targetInDifferrentSection = _.find(fieldRule.targetField, function(targetField) {
+            var targetSection = fieldSectionMapping[targetField.toString()];
+            var targetSectionId = targetSection && targetSection._id.toString();
+            return targetSectionId !== sectionIds[0];
+          });
+          if (targetInDifferrentSection) {
+            rulesFlaggedForDeletion[fieldRule._id.toString()] = true;
+          }
+        }
+      });
+    }
+  }
+
+  function updateOrDeleteRules(cb) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Function has a complexity of 29 which is greater than 20 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)